### PR TITLE
User documentation

### PR DESCRIPTION
**Pull request by alekseysemenov**
@ctapobep

### CODE REVIEW

File: documentation/user_doc.md [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+Basically it POST an HTTP requests from master to peers instance when new artifact is saved then peers invoke artifact pooling into proxy repository.
+
+## Setting up plugin
+
+Plugin should be installed into master and peer Nexus servers both.
+
+### Master instance
+
+1. Download source from [https://github.com/griddynamics/nexus-replication-plugin](https://github.com/griddynamics/nexus-replication-plugin).
+2. Run `mvm package` in root directory (project requires Apache Maven 3.0.4 -- 3.1 and JDK 1.7+).
```
Comment (kevstigneev) [inline documentation/user_doc.md]: mv**m** -> mv**n**

Maven version - anything below 3.2.1 is considered outdated nowadays. It should work on Maven 3.2.1 and higher.

File: documentation/user_doc.md [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+3. Unzip the `target/nexus-replication-plugin-1.0-SNAPSHOT-bundle.zip` file into the plugin-repository directory (located in `$NEXUS_HOME/sonatype-work/nexus/plugin-repository`).
```
Comment (kevstigneev) [inline documentation/user_doc.md]: I'd suggest to add `<finalName` directive to the POM so one won't need to take changing into account.

File: documentation/user_doc.md [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+4. Download a sample of the configuration file from [https://github.com/griddynamics/nexus-replication-plugin/blob/dev/replication-plugin.xml](https://github.com/griddynamics/nexus-replication-plugin/blob/dev/replication-plugin.xml).
```
Comment (kevstigneev) [inline documentation/user_doc.md]: dev -> master - `master` it the main branch.

File: documentation/user_doc.md [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+5. Copy the `replication-plugin.xml` file in the conf  directory (located in `$NEXUS_HOME/sonatype-work/nexus/conf`).
+6. Edit the config file to suit your needs (read instructions inside the file).
+7. Restart Nexus.
+
+### Peer instance
+
+You also should configure each Nexus peer instance. 
+
+1. Reproduce points 1 -- 3 from the previous list.
```
Comment (kevstigneev) [inline documentation/user_doc.md]: It seema m-dash (`--`) is not supported in GitHub markdown.

File: replication-plugin.xml [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+<server>
+    <url>http://localhost:8083/nexus</url>
+    <user>admin</user>
+    <password>admin123</password>
+</server>
+
+User and password are optional. If there is neither login nor password (or login is empty) server will use anonymous user.
+You could also define meUrl attribute for <configurations> tag tat determine URL of the Nexus instance.
+-->
+<configurations meUrl="http://localhost:8081/nexus">
```
Comment (kevstigneev) [inline replication-plugin.xml]: `meUrl` or `myUrl`?

File: documentation/user_doc.md [hunk 6]
```diff
@@ -0,0 +1,55 @@
+# Nexus Replication Plugin
+
+[https://github.com/griddynamics/nexus-replication-plugin](https://github.com/griddynamics/nexus-replication-plugin)
+
+This is Nexus plugin that provides replication logic when new (Maven) artifacts are stored in a hosted repository. 
+Basically it POST an HTTP requests from master to peers instance when new artifact is saved then peers invoke artifact pooling into proxy repository.
```
Comment (ctapobep) [inline documentation/user_doc.md]: I'd rephrase it like this: 
> Let's say you have 2 or more instances of Nexus. When artifact is uploaded to one of them, that Nexus will send notifications to its peers. Then the Receving Nexuses poll the artifact from the original instance.

File: documentation/user_doc.md [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+
+Plugin should be installed into master and peer Nexus servers both.
+
+### Master instance
+
+1. Download source from [https://github.com/griddynamics/nexus-replication-plugin](https://github.com/griddynamics/nexus-replication-plugin).
+2. Run `mvm package` in root directory (project requires Apache Maven 3.0.4 -- 3.1 and JDK 1.7+).
+3. Unzip the `target/nexus-replication-plugin-1.0-SNAPSHOT-bundle.zip` file into the plugin-repository directory (located in `$NEXUS_HOME/sonatype-work/nexus/plugin-repository`).
+4. Download a sample of the configuration file from [https://github.com/griddynamics/nexus-replication-plugin/blob/dev/replication-plugin.xml](https://github.com/griddynamics/nexus-replication-plugin/blob/dev/replication-plugin.xml).
+5. Copy the `replication-plugin.xml` file in the conf  directory (located in `$NEXUS_HOME/sonatype-work/nexus/conf`).
```
Comment (ctapobep) [inline documentation/user_doc.md]: in -> to (or into).

File: documentation/user_doc.md [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+6. Edit the config file to suit your needs (read instructions inside the file).
+7. Restart Nexus.
+
+### Peer instance
+
+You also should configure each Nexus peer instance.
```
Comment (ctapobep) [inline documentation/user_doc.md]: ... Otherwise they won't poll the uploaded artifact.

File: documentation/user_doc.md [hunk 9]
```diff
@@ -0,0 +1,55 @@
+# Nexus Replication Plugin
+
+[https://github.com/griddynamics/nexus-replication-plugin](https://github.com/griddynamics/nexus-replication-plugin)
+
+This is Nexus plugin that provides replication logic when new (Maven) artifacts are stored in a hosted repository. 
+Basically it POST an HTTP requests from master to peers instance when new artifact is saved then peers invoke artifact pooling into proxy repository.
+
+## Setting up plugin
+
+Plugin should be installed into master and peer Nexus servers both.
```
Comment (ctapobep) [inline documentation/user_doc.md]: Since for plugin there is no such thing as Master instance, I'd mention:
> You may have multiple Nexus instances configured as Masters. They will poll artifacts from each other.

I'm not sure whether Master is even a correct term here..

File: documentation/user_doc.md [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+
+You also should configure each Nexus peer instance. 
+
+1. Reproduce points 1 -- 3 from the previous list.
+2. Create proxy repository for the each hosted repository (repositories that should be replicated and located in the master Nexus). 
+4. Restart Nexus.
+
+## Plugin REST API
+
+Plugin REST API is available from `service/local/artifact/maven/update`. This resource receives POST request with a body like this:
```
Comment (ctapobep) [inline documentation/user_doc.md]: from -> at

File: replication-plugin.xml [hunk 11]
```diff
@@ -0,0 +1,27 @@
+<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
+<!--
+Nexus Replication Plugin - https://github.com/griddynamics/nexus-replication-plugin
+Configuration file - Should be located in $NEXUS_HOME/sonatype-work/nexus/conf/replication-plugin.xml
+Format : XML
```
Comment (ctapobep) [inline replication-plugin.xml]: I think the format is self-explanatory :)

File: replication-plugin.xml [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+
+You can configure one or more peer servers here.
+Each server configuration should be places into <servers></servers> tag.
```
Comment (ctapobep) [inline replication-plugin.xml]: be places -> be placed

File: replication-plugin.xml [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+Server configuration example:
+<server>
+    <url>http://localhost:8083/nexus</url>
+    <user>admin</user>
+    <password>admin123</password>
+</server>
+
+User and password are optional. If there is neither login nor password (or login is empty) server will use anonymous user.
+You could also define meUrl attribute for <configurations> tag tat determine URL of the Nexus instance.
```
Comment (ctapobep) [inline replication-plugin.xml]: tag tat determine -> tag that defines

File: documentation/user_doc.md (hunk 1 above)
Comment (ctapobep) [inline documentation/user_doc.md]: Hm.. can't we put the plugin to GD Nexus? Or to GitHub Releases?

File: documentation/contributor_doc.md [hunk 14]
```diff
@@ -0,0 +1,57 @@
+# Nexus Replication Plugin
+
+[https://github.com/griddynamics/nexus-replication-plugin](https://github.com/griddynamics/nexus-replication-plugin)
+
+This is Nexus plugin that provides replication logic when new (Maven) artifacts are stored in a hosted repository.
```
Comment (kevstigneev) [inline documentation/contributor_doc.md]: This plugin does _not_ provide replication logic. Replication is provided by Nexus core. The plugin just watch for uploads and notifies peers, reveives notification and triggers download to proxy repositories.

**Comment by kevstigneev** [comment]
Commit message. Few commits have the same not-so descriptive messages.
Follow recommendation in ProGit book or adhere to the guideline - https://docs.google.com/a/griddynamics.com/document/d/1Nz1kSWRPpmDiphVKgWwk85QDlK_1zOR8y1Cwc_JoW9c/edit?usp=sharing

**README.md**

- The plugin is for **Sonatype** Nexus, but it isn't mentioned. Just "Nexus" is too vague.
- "To be done" section should be updated or moved to Issues.

File: documentation/contributor_doc.md (hunk 14 above)
Comment (ctapobep) [inline documentation/contributor_doc.md]: Sounds debatable to me. I would say that the logic of the replication is provided by the plugin - of course some stuff is done by Nexus Core, but not the replication itself.. 

Or at least I'd say that it _facilitates_ the replication.

File: replication-plugin.xml [hunk 15]
```diff
@@ -0,0 +1,23 @@
...
+Each server configuration should be placed within <servers></servers> tag.
+Server configuration example:
+<server>
+    <url>http://localhost:8083/nexus</url>
+    <user>admin</user>
+    <password>admin123</password>
+</server>
+
+User and password are optional. If there is neither login nor password (or login is empty) server will use anonymous user.
+You could also define myUrl attribute for <configurations> tag that defines URL of the Nexus instance.
```
Comment (ctapobep) [inline replication-plugin.xml]: I just thought.. It will probably be useful to explain _why_ we need this _myUrl_ and how it's used:
- At least it will be clear for users
- Since Nexus may have different URLs, users may not understand what particular URL is expected here. So the doc will serve as a hint for them.

**Comment by ctapobep** [comment]
BTW, why do we use `dev` branch here?

**Comment by kevstigneev** [comment]
@alekseysemenov can you re-submit the PR to "master" branch?

The docs definitely need some work, but they're a good starting point. So I suggets to accept it to "master".
